internal/staging/start_testrunner.go

```diff
@@ -32,68 +37,81 @@ func (c *controller) startTesting(queue *s2hv1.Queue) error {
 	skipTest, testRunners, err := c.checkTestConfig(queue)
-	if err != nil {
+	if err != nil || skipTest {
 		return err
-	} else if skipTest {
-		return nil
 	}
 
 	// trigger the tests
-	for _, testRunner := range testRunners {
-		if err := c.triggerTest(queue, testRunner); err != nil {
+	notTriggeredTest := !queue.Status.IsConditionTrue(s2hv1.QueueTestTriggered)
+	notReachTriggerTimeout := queue.Status.StartTestingTime != nil &&
+		metav1.Now().Sub(queue.Status.StartTestingTime.Time) <= testTriggerTimeout
+	// check testing timeout
+	if notTriggeredTest && notReachTriggerTimeout {
+		err = c.triggerTest(queue, testRunners)
+		if err != nil {
+			// retry util time passed testTriggerTimeout
 			return err
 		}
 	}
 
-	if !queue.Status.IsConditionTrue(s2hv1.QueueTestTriggered) {
-		queue.Status.SetCondition(
-			s2hv1.QueueTestTriggered,
-			v1.ConditionTrue,
-			"queue testing triggered")
-
-		// update queue back to k8s
-		if err := c.updateQueue(queue); err != nil {
-			return err
-		}
+	// set state, test has been triggered
+	queue.Status.SetCondition(
+		s2hv1.QueueTestTriggered,
+		v1.ConditionTrue,
+		"queue testing triggered")
+	// update queue back to k8s
+	if err := c.updateQueue(queue); err != nil {
+		return err
 	}
 
 	// get result from tests (polling check)
-	finished := true
 	testCondition := v1.ConditionTrue
 	message := "queue testing succeeded"
+	allTestFinished := true
+	triggerTestMsg := ""
 	for _, testRunner := range testRunners {
 		testRunnerName := testRunner.GetName()
-		testResult, err := c.getTestResult(queue, testRunner)
-		if err != nil {
-			return err
+
+		if !testRunner.IsTriggered(queue) {
+			// add message in k8s object
+			triggerTestMsg += fmt.Sprintf("cannot trigger test on %s ", testRunnerName)
+			continue
 		}
 
-		switch testResult {
-		case testResultUnknown:
-			finished = false
-		case testResultFailure, testResultSuccess:
-			if testResult == testResultFailure {
-				testCondition = v1.ConditionFalse
-				message = "queue testing failed"
-			}
+		testResult, err := c.getTestResult(queue, testRunner)
 
-			if err := c.setTestResultCondition(queue, testRunnerName, testResult); err != nil {
-				return err
-			}
+		// unfinished test
+		if testResult == testResultUnknown && err == nil {
+			allTestFinished = false
+			continue
 		}
-	}
 
-	if finished {
-		if err := c.updateTestQueueCondition(queue, testCondition, message); err != nil {
+		// if finished, then update test result
+		if err := c.setTestResultCondition(queue, testRunnerName, testResult); err != nil {
 			return err
 		}
-	}
 
+		// if some runners test were failed
+		if testResult == testResultUnknown || testResult == testResultFailure {
+			testCondition = v1.ConditionFalse
+			message = "queue testing failed"
+			if triggerTestMsg != "" {
+				message = triggerTestMsg
+			}
+		}
+	}
+	// test finished, change state to `s2hv1.Collecting`
+	if allTestFinished {
+		return c.updateTestQueueCondition(queue, testCondition, message)
```

Comment:
in case of trigger test runner timeout, would the queue condition `QueueTestTriggered` be set to `false` before changing state to collecting?

Reply:
It will changed to `false` kub